storage/store.go

```diff
@@ -175,42 +193,27 @@ func (e *NotBootstrappedError) Error() string {
 	return "store has not been bootstrapped"
 }
 
-// storeRangeIterator is an implementation of rangeIterator which
-// cycles through a store's rangesByKey slice.
-type storeRangeIterator struct {
-	store     *Store
-	remaining int
-	index     int
-}
-
-func newStoreRangeIterator(store *Store) *storeRangeIterator {
-	r := &storeRangeIterator{
-		store: store,
-	}
-	r.Reset()
-	return r
+// storeRangeSet is an implementation of rangeSet which
+// cycles through a store's rangesByKey btree with mutex locking.
+type storeRangeSet struct {
+	store *Store
 }
 
-func (si *storeRangeIterator) Next() *Range {
-	si.store.mu.Lock()
-	defer si.store.mu.Unlock()
-	if index, remaining := si.index, len(si.store.rangesByKey)-si.index; remaining > 0 {
-		si.index++
-		si.remaining = remaining - 1
-		return si.store.rangesByKey[index]
-	}
-	return nil
-}
-
-func (si *storeRangeIterator) EstimatedCount() int {
-	return si.remaining
+func (rs *storeRangeSet) Visit(iterator func(*Range) bool) {
+	// Hold the lock except when the iterator is called.
```

Comment:
Copying pointers to ranges is dangerous since we may hold on to a range that has been removed. I think I would make a copy of the raft ids instead, and then we can look them up to make sure they're still there when their turn comes. (or maybe we just need to make sure we're robust about removed ranges in the scanner itself. if we don't hold the lock while the visitor runs then it may be removed during the visitor).

Either way, I think it's better to copy everything from the btree into a slice than to risk uncertain concurrency semantics. @spencerkimball?